_posts/2019-06-27-list-strategy.md

```diff
@@ -0,0 +1,95 @@
+---
+layout: posts
+title: "Automatic execution strategy selection in Bazel 0.27"
+authors:
+  - ishikhman
+---
+**tl;dr**: Exciting news! After 0.27 release Bazel will *auto select* a suitable execution strategy if none provided. Customization and strategy enforcement would still be possible and will be much easier to configure.
+
+## How it was before bazel 0.27 release
+
+It was possible to configure Bazel's execution strategy for a build via flags `--spawn_strategy=`, `--strategy=Mnemonic=` and `--strategy_regexp=Regex=`. This mechanism was quite powerful and widely used, but had some drawbacks:
+
+- Bazel's defaults did not account for remote execution at all, therefore we had to make sure to set the execution strategy for every mnemonic manually.
+
+- Starlark rules providing a persistent worker had to ship a .bazelrc file that sets --strategy=Mnemonic=worker (i.e. look at rules_scala). 
+
+- Bazel has hardcoded defaults for native action mnemonics. For example, Bazel will use sandboxing (if it's available) for all actions by default but has a hardcoded default to use persistent workers for Java compilation actions.
+
+- To configure fallback strategy one had to provide additional flag `--remote_local_fallback_strategy=`.
+
+All of the above are now unnecessary with this change and the default behaviour is much more reliable and predictable.
+
+### Example
+I want to configure a build to run remotely and fallback onto `local` strategy in case remote execution is not possible. The configuration would be something like:
+
+``` 
+$ bazel build
+ --spawn_strategy=remote
+ --strategy=Javac=remote
+ --strategy=Closure=remote
+ --strategy=Genrule=remote
+ --remote_local_fallback_strategy=local
+ ... 
+```
+
+
+## How it is after bazel 0.27
+- Bazel now *auto-detects* the execution strategy, if no strategy flag provided.  
+If none of the strategy flags was used, bazel will generate a default list of strategies `remote,worker,sandboxed,local` and, for every action it wants to execute, will pick up the first strategy that can execute it.
+
+- The user can pass comma-separated lists of strategies to the above mentioned flags: `--spawn_strategy=remote,worker,linux-sandbox`.  
```

Comment:
pass **a** comma-separated **list** of strategies

Reply:
done